File: src/modules/content.js
/**
 * Pure content manipulation functions.
 * All functions return values instead of mutating state.
 */

function processTemplate(template) {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');

  return template.replace(/\{\{date:YYYY-MM-DD\}\}/g, `${year}-${month}-${day}`);
}

function findSectionIndex(lines, sectionName) {
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].startsWith('## ') && lines[i].includes(sectionName)) {
      return i;
    }
  }
  return -1;
}

/**
 * @returns {{ newContent: string, insertedLine: number }} | null (null means section not found)
 */
function addToSection(currentContent, sectionName, content) {
  const lines = currentContent.split('\n');
  const sectionIndex = findSectionIndex(lines, sectionName);

  if (sectionIndex === -1) {
    return null;
  }

  let insertIndex = sectionIndex + 1;
  let lastContentLine = sectionIndex;
  let hasContent = false;

  while (insertIndex < lines.length && !lines[insertIndex].startsWith('## ')) {
    if (lines[insertIndex].trim() !== '') {
      lastContentLine = insertIndex;
      hasContent = true;
    }
    insertIndex++;
  }

  let actualInsertLine;
  if (!hasContent) {
    lines.splice(sectionIndex + 1, 0, content);
    actualInsertLine = sectionIndex + 1;
  } else {
    lines.splice(lastContentLine + 1, 0, content);
    actualInsertLine = lastContentLine + 1;
  }

  return {
    newContent: lines.join('\n'),
    insertedLine: actualInsertLine + 1
  };
}

/**
 * @returns {{ newContent: string, insertedLine: number }}
 */
function addContent(currentContent, newContent, mode = 'append') {
  const lines = currentContent.split('\n');
  let resultContent;
  let insertLine;

  switch (mode) {
  case 'append':
    resultContent = currentContent + '\n' + newContent;
    insertLine = lines.length + 1;
    break;
  case 'prepend':
    resultContent = newContent + '\n' + currentContent;
    insertLine = 1;
    break;
  case 'replace':
    resultContent = newContent;
    insertLine = 1;
    break;
  }

  return { newContent: resultContent, insertedLine: insertLine };
}

/**
 * @returns {{ newContent: string, insertedLine: number }} | null
 */
function insertContentAtLine(currentContent, content, lineNumber) {
  const lines = currentContent.split('\n');

  if (lineNumber >= 0 && lineNumber <= lines.length) {
    lines.splice(lineNumber, 0, content);
    return { newContent: lines.join('\n'), insertedLine: lineNumber + 1 };
  }

  return null;
}

/**
 * @returns {{ newContent: string }} | null
 */
function replaceContentAtLine(currentContent, content, lineNumber) {
  const lines = currentContent.split('\n');

  if (lineNumber > 0 && lineNumber <= lines.length) {
    lines[lineNumber - 1] = content;
    return { newContent: lines.join('\n') };
  }

  return null;
}

/**
 * Parse input prefix to determine section routing.
 * @returns {{ section: string, formattedContent: string, logType: string, rawContent: string }} | null
 */
function parseContentInput(input) {
  const trimmed = input.trim();

  if (trimmed.startsWith('[]')) {
    const rawContent = trimmed.slice(2).trim();
    return { section: 'Tasks', formattedContent: `- [ ] ${rawContent}`, logType: 'task', rawContent };
  } else if (trimmed.startsWith('-')) {
    const rawContent = trimmed.slice(1).trim();
    return { section: 'Ideas', formattedContent: `- ${rawContent}`, logType: 'idea', rawContent };
  } else if (trimmed.startsWith('?')) {
    const rawContent = trimmed.slice(1).trim();
    return { section: 'Questions', formattedContent: `- ${rawContent}`, logType: 'question', rawContent };
  } else if (trimmed.startsWith('!')) {
    const rawContent = trimmed.slice(1).trim();
    return { section: 'Insights', formattedContent: `- ${rawContent}`, logType: 'insight', rawContent };
  }

  return null;
}

/**
 * Inject or update metadata (updated_at) in content.
 */
function injectMetadata(content) {
  const lines = content.split('\n');
  const hasMetadata = lines.some(line => line.includes('updated_at:'));

  if (!hasMetadata && lines.length > 0) {
    const now = new Date();
    const metadata = [
      '---',
      `updated_at: ${now.toISOString()}`,
      '---'
    ];

    if (lines[0].startsWith('#')) {
      lines.splice(1, 0, ...metadata);
      return lines.join('\n');
    }
  } else if (hasMetadata) {
    const updatedLines = lines.map(line => {
      if (line.includes('updated_at:')) {
        return `updated_at: ${new Date().toISOString()}`;
      }
      return line;
    });
    return updatedLines.join('\n');
  }

  return content;
}

module.exports = {
  processTemplate,
  findSectionIndex,
  addToSection,
  addContent,
  insertContentAtLine,
  replaceContentAtLine,
  parseContentInput,
  injectMetadata
};

File: internal/content/content.go
package content

import (
	"fmt"
	"regexp"
	"strings"
)

// ContentResult holds the result of a content manipulation operation.
type ContentResult struct {
	NewContent   string
	InsertedLine int // 1-based line number
}

// ParsedInput holds the result of parsing a prefixed user input.
type ParsedInput struct {
	Section          string // "Tasks", "Ideas", "Questions", "Insights"
	FormattedContent string // "- [ ] Buy groceries"
	LogType          string // "task", "idea", "question", "insight"
	RawContent       string // "Buy groceries"
}

// FindSectionIndex returns the 0-based index of the "## sectionName" header, or -1.
func FindSectionIndex(lines []string, sectionName string) int {
	for i, line := range lines {
		if strings.HasPrefix(line, "## ") && strings.Contains(line, sectionName) {
			return i
		}
	}
	return -1
}

// AddToSection inserts content after the last non-empty line in the named section.
// Returns nil if the section is not found.
func AddToSection(currentContent, sectionName, content string) *ContentResult {
	lines := strings.Split(currentContent, "\n")
	sectionIndex := FindSectionIndex(lines, sectionName)

	if sectionIndex == -1 {
		return nil
	}

	insertIndex := sectionIndex + 1
	lastContentLine := sectionIndex
	hasContent := false

	for insertIndex < len(lines) && !strings.HasPrefix(lines[insertIndex], "## ") {
		if strings.TrimSpace(lines[insertIndex]) != "" {
			lastContentLine = insertIndex
			hasContent = true
		}
		insertIndex++
	}

	var actualInsertLine int
	if !hasContent {
		// Insert right after section header
		lines = insertAt(lines, sectionIndex+1, content)
		actualInsertLine = sectionIndex + 1
	} else {
		// Insert after last content line
		lines = insertAt(lines, lastContentLine+1, content)
		actualInsertLine = lastContentLine + 1
	}

	return &ContentResult{
		NewContent:   strings.Join(lines, "\n"),
		InsertedLine: actualInsertLine + 1, // 1-based
	}
}

// AddContent appends, prepends, or replaces content.
func AddContent(currentContent, newContent, mode string) *ContentResult {
	lines := strings.Split(currentContent, "\n")

	switch mode {
	case "prepend":
		return &ContentResult{
			NewContent:   newContent + "\n" + currentContent,
			InsertedLine: 1,
		}
	case "replace":
		return &ContentResult{
			NewContent:   newContent,
			InsertedLine: 1,
		}
	default: // "append"
		return &ContentResult{
			NewContent:   currentContent + "\n" + newContent,
			InsertedLine: len(lines) + 1,
		}
	}
}

// InsertContentAtLine inserts content at the given 0-based line number.
// Returns nil if lineNumber is out of bounds.
func InsertContentAtLine(currentContent, content string, lineNumber int) *ContentResult {
	lines := strings.Split(currentContent, "\n")

	if lineNumber < 0 || lineNumber > len(lines) {
		return nil
	}

	lines = insertAt(lines, lineNumber, content)
	return &ContentResult{
		NewContent:   strings.Join(lines, "\n"),
		InsertedLine: lineNumber + 1,
	}
}

// ReplaceContentAtLine replaces the line at the given 1-based line number.
// Returns nil if lineNumber is out of bounds.
func ReplaceContentAtLine(currentContent, content string, lineNumber int) *ContentResult {
	lines := strings.Split(currentContent, "\n")

	if lineNumber < 1 || lineNumber > len(lines) {
		return nil
	}

	lines[lineNumber-1] = content
	return &ContentResult{
		NewContent: strings.Join(lines, "\n"),
	}
}

// ParseContentInput parses prefixed input into a structured result.
// Returns nil if the input doesn't match any known prefix.
func ParseContentInput(input string) *ParsedInput {
	trimmed := strings.TrimSpace(input)

	if strings.HasPrefix(trimmed, "[]") {
		raw := strings.TrimSpace(trimmed[2:])
		return &ParsedInput{
			Section:          "Tasks",
			FormattedContent: fmt.Sprintf("- [ ] %s", raw),
			LogType:          "task",
			RawContent:       raw,
		}
	}

	if strings.HasPrefix(trimmed, "-") {
		raw := strings.TrimSpace(trimmed[1:])
		return &ParsedInput{
			Section:          "Ideas",
			FormattedContent: fmt.Sprintf("- %s", raw),
			LogType:          "idea",
			RawContent:       raw,
		}
	}

	if strings.HasPrefix(trimmed, "?") {
		raw := strings.TrimSpace(trimmed[1:])
		return &ParsedInput{
			Section:          "Questions",
			FormattedContent: fmt.Sprintf("- %s", raw),
			LogType:          "question",
			RawContent:       raw,
		}
	}

	if strings.HasPrefix(trimmed, "!") {
		raw := strings.TrimSpace(trimmed[1:])
		return &ParsedInput{
			Section:          "Insights",
			FormattedContent: fmt.Sprintf("- %s", raw),
			LogType:          "insight",
			RawContent:       raw,
		}
	}

	return nil
}

func insertAt(lines []string, index int, value string) []string {
	result := make([]string, len(lines)+1)
	copy(result, lines[:index])
	result[index] = value
	copy(result[index+1:], lines[index:])
	return result
}

var dateTemplateRe = regexp.MustCompile(`\{\{date:YYYY-MM-DD\}\}`)
